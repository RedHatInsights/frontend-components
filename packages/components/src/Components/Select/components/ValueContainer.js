import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button } from '@patternfly/react-core';

class ValueContainer extends Component {
    state = {
        showAll: false,
    }
    render() {
        const { isMulti, ...props } = this.props;
        const { showAll } = this.state;
        if (isMulti && props.children) {
            return (
                <div className="ins-c-select__value--container">
                    {
                        showAll ?
                        props.children[0] :
                        props.children[0] && props.children[0][0] ? props.children[0][0] : props.children[0]
                    }
                    {props.children[0] && props.children[0].length > 1 && (
                        <Button
                            className="ins-c-select__value--container-chipgroup"
                            onClick={() => this.setState(({ showAll }) => ({ showAll: !showAll }))}
                            variant="plain"
                        >
                            <span>{
                                showAll ?
                                props.selectProps.showLessLabel :
                                `${props.children[0].length - 1} ${props.selectProps.showMoreLabel}`
                            }</span>
                        </Button>
                    )}
                    {
                        Array.isArray(props.children) ? props.children[1] && props.children[1] : props.children
                    }
                </div>
            );
        }

        return props.children;
    }
}

ValueContainer.propTypes = {
    isMulti: PropTypes.bool,
    getStyles: PropTypes.func.isRequired,
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.arrayOf(PropTypes.node)]).isRequired,
    selectProps: PropTypes.shape({
        showLessLabel: PropTypes.string,
        showMoreLabel: PropTypes.string
    })
};

ValueContainer.defaultProps = {
    isMulti: false,
    selectProps: {
        showLessLabel: 'Show less',
        showMoreLabel: 'more'
    }
}

export default ValueContainer;
