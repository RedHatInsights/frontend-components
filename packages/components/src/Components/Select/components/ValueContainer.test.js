import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import ValueContainer from './ValueContainer';

const requiredProps = {
    getStyles: () => undefined,
    cx: () => undefined
};

describe('ValueContainer component', () => {
    describe('should render', () => {
        it('required props', () => {
            const wrapper = render(<ValueContainer {...requiredProps}>
                <div>aaa</div>
            </ValueContainer>);
            expect(toJson(wrapper)).toMatchSnapshot();
        });

        describe('isMulti', () => {
            it('required props', () => {
                const wrapper = render(<ValueContainer isMulti {...requiredProps}>
                    <div>aaa</div>
                </ValueContainer>);
                expect(toJson(wrapper)).toMatchSnapshot();
            });

            it('two children', () => {
                const wrapper = render(<ValueContainer isMulti {...requiredProps}>
                    <div>aaa</div>
                    <div>aaa</div>
                </ValueContainer>);
                expect(toJson(wrapper)).toMatchSnapshot();
            });

            it('show more button', () => {
                const wrapper = render(<ValueContainer isMulti {...requiredProps}>
                    {[ 'first', 'second'].map(item => item)}
                    <div>aaa</div>
                </ValueContainer>);
                expect(toJson(wrapper)).toMatchSnapshot();
            });
        });
    });

    describe('api', () => {
        it('should toggle', () => {
            const wrapper = mount(<ValueContainer isMulti {...requiredProps}>
                {['first', 'second'].map(item => item)}
                <div>aaa</div>
            </ValueContainer>);
            wrapper.find('button.ins-c-select__value--container-chipgroup').first().simulate('click');
            wrapper.update();
            expect(toJson(wrapper)).toMatchSnapshot();
        });
    });
});
