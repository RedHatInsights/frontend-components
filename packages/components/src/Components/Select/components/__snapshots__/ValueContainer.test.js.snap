// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueContainer component api should toggle 1`] = `
<ValueContainer
  cx={[Function]}
  getStyles={[Function]}
  isMulti={true}
  selectProps={
    Object {
      "showLessLabel": "Show less",
      "showMoreLabel": "more",
    }
  }
>
  <div
    className="ins-c-select__value--container"
  >
    first
    second
    <Button
      className="ins-c-select__value--container-chipgroup"
      onClick={[Function]}
      variant="plain"
    >
      <button
        aria-disabled={null}
        aria-label={null}
        className="pf-c-button pf-m-plain ins-c-select__value--container-chipgroup"
        disabled={false}
        onClick={[Function]}
        tabIndex={null}
        type="button"
      >
        <span>
          Show less
        </span>
      </button>
    </Button>
    <div>
      aaa
    </div>
  </div>
</ValueContainer>
`;

exports[`ValueContainer component should render isMulti required props 1`] = `
<div
  class="ins-c-select__value--container"
>
  <div>
    aaa
  </div>
</div>
`;

exports[`ValueContainer component should render isMulti show more button 1`] = `
<div
  class="ins-c-select__value--container"
>
  first
  <button
    class="pf-c-button pf-m-plain ins-c-select__value--container-chipgroup"
    type="button"
  >
    <span>
      1 more
    </span>
  </button>
  <div>
    aaa
  </div>
</div>
`;

exports[`ValueContainer component should render isMulti two children 1`] = `
<div
  class="ins-c-select__value--container"
>
  <div>
    aaa
  </div>
  <div>
    aaa
  </div>
</div>
`;

exports[`ValueContainer component should render required props 1`] = `
<div>
  aaa
</div>
`;
